perfaware/sim86/shared/Sim8086.cpp

```diff
@@ -73,6 +110,85 @@ static void SimulateInstruction(const instruction& Instruction)
                 } break;
             }
         } break;
+        case Op_add:
+        case Op_sub:
+        case Op_cmp:
+        {
+            const instruction_operand& Dest = Instruction.Operands[0];
+            const instruction_operand& Source = Instruction.Operands[1];
+            u16 Value;
+            u16 Result;
+            switch (Source.Type)
+            {
+                case Operand_Register:
+                {
+                    Value = Registers[Source.Register.Index];
+                    if (Source.Register.Count == 1) // Accessing half registers
+                    {
+                        Value >>= 8 * Source.Register.Offset;
+                    }
+                } break;
+                case Operand_Immediate:
+                {
+                    Value = Source.Immediate.Value;
+                } break;
+            }
+
+            switch (Dest.Type)
+            {
+                case Operand_Register:
+                {
+                    
+                    if (Dest.Register.Count == 1) // Accessing half registers
+                    {
+                        u8* DestPointer = reinterpret_cast<u8*>(&Registers[Dest.Register.Index]) + Dest.Register.Offset;
+
+                        switch (Instruction.Op)
+                        {
+                            case Op_add:
+                            {
+                                Result = *DestPointer + Value;
+                                *DestPointer = static_cast<u8>(Result);
+                            } break;
+                            case Op_sub:
+                            {
+                                Result = *DestPointer - Value;
+                                *DestPointer = static_cast<u8>(Result);
+                            } break;
+                            case Op_cmp:
+                            {
+                                Result = *DestPointer - Value;
+                            } break;
+                        }
+                    }
+                    else
+                    {
+                        switch (Instruction.Op)
+                        {
+                            case Op_add:
+                            {
+                                Registers[Dest.Register.Index] += Value;
+                                Result = Registers[Dest.Register.Index];
+                            } break;
+                            case Op_sub:
+                            {
+                                Registers[Dest.Register.Index] -= Value;
+                                Result = Registers[Dest.Register.Index];
+                            } break;
+                            case Op_cmp:
+                            {
+                                Result = Registers[Dest.Register.Index] - Value;
+                            } break;
+                        }
+                        
+                    }
+                } break;
+            }
```

Comment:
First, you make a switch that handles add, sub and cmp in the same way, then inside you switch on the dest type, and then inside you switch on the op again, add sub or cmp. The benefit is that it saves you the need to write the code to get the source register multiple times, and it saves you the need to call set flags and print flags multiple times.

However, it causes an additional switch statement that can be avoided.

I'm not sure which way is better yet, probably need to let the code grow with more cases to see.

Reply:
I was on the fence about this one too and couldn't find a solution I was really happy with. 
We'll see what happens